packages/ui/src/components/Tooltip.tsx

```diff
@@ -0,0 +1,332 @@
+  stayOnScreen?: boolean;
+  show: boolean;
+  anchorRef: React.MutableRefObject<View | null>;
+  children: React.ReactNode;
+};
+
+const getPositions = ({
+  placement,
+  anchor,
+  tooltip,
+}: {
+  placement: string;
+  anchor: Size;
+  tooltip: Size;
+}): TooltipState => {
+  switch (placement) {
+    case "top":
+      return {
+        tooltipPosition: {
+          top: -tooltip.height - 16,
+          left: anchor.width / 2 - tooltip.width / 2,
+        },
+        trianglePosition: {
+          top: tooltip.height - 1,
+          left: tooltip.width / 2 - 7.5,
+        },
+      };
+    case "right":
+      return {
+        tooltipPosition: {
+          top: anchor.height / 2 - tooltip.height / 2,
+          left: anchor.width + 15,
+        },
+        trianglePosition: { top: tooltip.height / 2 - 7.5, left: -15 },
+      };
+    case "bottom":
+      return {
+        tooltipPosition: {
+          top: anchor.height + 15,
+          left: anchor.width / 2 - tooltip.width / 2,
+        },
+        trianglePosition: { top: -15, left: tooltip.width / 2 - 7.5 },
+      };
+    case "left":
+      return {
+        tooltipPosition: {
+          top: anchor.height / 2 - tooltip.height / 2,
+          left: -tooltip.width - 16,
+        },
+        trianglePosition: {
+          top: tooltip.height / 2 - 7.5,
+          left: tooltip.width - 1,
+        },
+      };
+    case "overlay":
+      return {
+        tooltipPosition: {
+          top: 0,
+          left: 0,
+          width: anchor.width,
+          height: anchor.height,
+        },
+        trianglePosition: {},
+      };
+    case "inset-bottom-left":
+      return {
+        tooltipPosition: { bottom: 0, left: 0 },
+        trianglePosition: {},
+      };
+    case "inset-top-left":
+      return {
+        tooltipPosition: { top: 0, left: 0 },
+        trianglePosition: {},
+      };
+    case "inset-top-right":
+      return {
+        tooltipPosition: { top: 0, right: 0 },
+        trianglePosition: {},
+      };
+    case "inset-bottom-right":
+      return {
+        tooltipPosition: { bottom: 0, right: 0 },
+        trianglePosition: {},
+      };
+    default:
+      return { tooltipPosition: {}, trianglePosition: {} };
+  }
+};
+
+type UIPosition = {
+  top?: number;
+  bottom?: number;
+  left?: number;
+  right?: number;
+  width?: number;
+  height?: number;
+};
+
+type TooltipState = {
+  tooltipPosition?: UIPosition;
+  trianglePosition?: UIPosition;
+};
+
+type Position = {
+  x: number;
+  y: number;
+};
+
+type Size = {
+  width: number;
+  height: number;
+};
+
+export default (props: TooltipProps) => {
+  const { placement = "top", show, anchorRef, children } = props;
+
+  const tooltipRef = useRef<View | null>(null);
+  const containerRef = useRef<View | null>(null);
+
+  const [{ tooltipPosition, trianglePosition }, dispatch] = useReducer(
+    (
+      state: TooltipState,
+      { type, payload }: { type: string; payload: any },
```

Comment:
Ah, this `any`! Exactly what I try to avoid by using TypeScript. The trouble here is that if I change the argument to `getPositions` (e.g. I thought I might rename `anchor` to `anchorSize` to be clearer), now this code will fail at runtime but won't generate any errors at build time! That's no good.

One alternative is to specify the types specifically: `{type: "init", payload: {placement: string, /* etc */}} | {type: "move", payload: Position}`. But rather than make the change myself, I thought I'd ask: might it be more straightforward to inline the state updates into the effects below? Each command is only used from one place, so I'm not sure `useReducer` is buying us anything here, and it costs us locality. I'm OK either way, but I do want to ditch the `any`!

Reply:
Yeah typescript + `useReducer` felt particularly awkward. This was more a result of my default habit of useReducer over useState. I don't think it's necessary here.